cogent/src/Cogent/Parser.hs

```diff
@@ -480,19 +492,30 @@ monotype = do avoidInitial
     fList = (Just . (:[])) <$> identifier
         <|> parens ((reservedOp ".." >> return Nothing) <|> (commaSep identifier >>= return . Just))
 
-polytype = PT <$ reserved "all" <*> (((:[]) <$> kindSignature) <|> parens (commaSep1 kindSignature)) <* reservedOp "." <*> monotype
-       <|> PT [] <$> monotype
-
-kindSignature = do n <- variableName
-                   (n,) <$> (reservedOp ":<" *> (kind <?> "kind")
-                          <|> pure (K False False False))
+polytype = polytype' <|> PT [] [] <$> monotype
+  where
+    polytype' = do
+      reserved "all"
+      hs <- ((:[]) <$> klSignature) <|> parens (commaSep1 klSignature)
+      reservedOp "."
+      t <- monotype
+      return $ PT (hs >>= flt1) (hs >>= flt2) t
+    flt1 x | Left v <- snd x  = pure (fst x, v)
+           | otherwise        = mempty
+    flt2 x | Right v <- snd x = pure (fst x, v)
+           | otherwise        = mempty
```

Comment:
Might be better to write as:
```
    flt1 (x,y) | Left v <- y  = pure (x, v)
               | otherwise    = mempty
    flt2 (x,y) | Right v <- y = pure (x, v)
               | otherwise    = mempty
```

Reply:
sure, will change in next commit